### PR TITLE
ADPCM decoder

### PR DESCRIPTION
**Pull request by trumank**
This should fix all the ADPCM audio problems. I think there are still some problems with the actual decoder, but I'm not sure. It seems like the audio is more quiet that it should be and sometime the output buffer is longer than it should be. In practice, I have not noticed a lack of quality in the output as one might expect with such issues. Any thoughts?

### CODE REVIEW

**Comment by tjvr** [comment]
For #260

**Comment by tjvr** [comment]
This is great, thanks so much @trumank! :D I ran through all the projects linked off #260, and I can't find a project where the sound doesn't work, which is awesome.

The audio sounds fine to me. It's definitely quieter, but that seems to be to with how Phosphorus plays audio/uses the volume, and not your decoder. For example, compare this project which does not need the decoder: <http://phosphorus.github.io/#88929725> vs <https://scratch.mit.edu/projects/88929725/>. If anything the quality of the Phosphorus audio sounds better to me...!

I see the following error in the console, should/can we silence this?

> Uncaught (in promise) DOMException: Unable to decode audio data

**Comment by nathan** [comment]
This is awesome! I'm going to have to review this one more thoroughly than bug fix/optimization PRs, so sorry if it takes me a few days to get to it.

**Comment by trumank** [comment]
> Uncaught (in promise) DOMException: Unable to decode audio data

This can be silenced by checking if the file is ADPCM encoded before trying to open it with the native decoder and letting it fail. Is it worth it?

**Comment by tjvr** [comment]
> This can be silenced by checking if the file is ADPCM encoded before trying to open it with the native decoder and letting it fail

Ah, I see; it's not a case of simply adding a try/catch. I'd say not worth fixing. :-)

File: phosphorus.js [hunk 1]
```diff
@@ -418,14 +421,94 @@ var P = (function() {
...
+          if (sample < -32768) sample = -32768;
+          channel[j++] = sample / 32768;
+        }
+      }
+      return cb(buffer);
+    } else {
+      console.warn('Unrecognized WAV format ' + format);
+      return cb(null);
+    }
+    cb(null);
```
Comment (tjvr) [inline phosphorus.js]: Why is the callback returned in every case but this one?

Comment (trumank) [inline phosphorus.js]: Because it's already at the end. :P

Comment (trumank) [inline phosphorus.js]: Although I do see now that the code could look a little nicer. I'll move a few things around.

**Comment by tjvr** [comment]
Unfortunately this doesn't seem to work in Firefox :(

> The buffer passed to decodeAudioData contains invalid content which cannot be decoded successfully.
> RangeError: argument 1 accesses an index that is out of range phosphorus.js:433:9
> EncodingError: The given encoding is not supported.

**Comment by tjvr** [comment]
Here's an [example project](https://scratch.mit.edu/projects/3120195/) that seems happy in Chrome, but sad in Firefox.

**Comment by tjvr** [comment]
@nathan Thanks for having a look at phosphorus today! What's the status of this PR? Even merging it as-is and improving it later would be great. :)

**Comment by nathan** [comment]
LGTM. Thanks, @trumank.

I think I stretched the definition of "a few" a bit too far.

**Comment by trumank** [comment]
Glad to see this finally merged!

**Pull request merged**